lib/sources/bulk_assembly.rb

```diff
@@ -0,0 +1,130 @@
+module Sources
+  class BulkAssembly
+    attr_reader :output
+
+    def initialize(params)
+      @sources = params[:sources]
+      @endpoints = params[:endpoints]
+      @applications = params[:applications]
+
+      # separate out the superkey authentications.
+      @superkeys, @authentications = params[:authentications]&.partition { |auth| auth[:authtype] == "superkey" }
+    end
+
+    def process
+      Source.transaction do
+        @output = {}.tap do |output|
+          # Create the base source(s)
+          output[:sources] = create_sources(@sources)
+
+          # Create the superkey authentications (if there are any)
+          output[:authentications] = create_authentications(@superkeys, output)
+
+          # Create the endpoints
+          output[:endpoints] = create_endpoints(@endpoints, output)
+
+          # Create the applications (sends superkey requests via callback)
+          output[:applications] = create_applications(@applications, output)
+
+          # Create the remaining authentications that aren't superkeys.
+          if output[:authentications]
+            output[:authentications].concat(create_authentications(@authentications, output))
+          else
+            output[:authentications] = create_authentications(@authentications, output)
+          end
+        end
+
+        self
+      rescue => e
+        Rails.logger.error("Error bulk processing from payload: Sources: #{@sources}, Endpoints: #{@endpoints}, Applications: #{@applications}, Authentications: #{@authentications}. Error: #{e}")
+
+        raise
+      end
+    end
+
+    def create_sources(sources)
+      sources&.map do |source|
+        # get the source type by ID or by type string
+        srct = SourceType.find_by(:id => source.delete(:source_type_id)) || SourceType.find_by(:name => source.delete(:type))
+
+        raise ActiveRecord::ActiveRecordError, "Source Type not found" if srct.nil?
+
+        Source.create!(source.merge!(:source_type => srct))
+      end
+    end
+
+    def create_endpoints(endpoints, resources)
+      endpoints&.map do |endpoint|
+        src = find_resource(resources, :sources, endpoint.delete(:source_name))
```

Comment:
what is the structure of payload? This column doesn't exist in the model so calling this operation needs undocumented knowledge

Reply:
The structure is in the openapi.json, here: https://github.com/lindgrenj6/sources-api/blob/932ba4c3f5f45e20a8265742bd04e5fca6893018/public/doc/openapi-3-v3.1.json#L2234-L2348